tests/usbredir_test.go

```diff
@@ -0,0 +1,177 @@
+
+	var err error
+	var virtClient kubecli.KubevirtClient
+	var vmi *v1.VirtualMachineInstance
+
+	Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component] A VirtualMachineInstance without usbredir support", func() {
+		tests.BeforeAll(func() {
+			virtClient, err = kubecli.GetKubevirtClient()
+			util.PanicOnError(err)
+
+			tests.BeforeTestCleanup()
+			vmi = tests.NewRandomVMI()
+			Expect(virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(util.NamespaceTestDefault).Body(vmi).Do(context.Background()).Error()).To(Succeed())
+			tests.WaitForSuccessfulVMIStart(vmi)
+		})
+
+		FIt("should fail to connect to VMI's usbredir socket", func() {
+			usbredirVMI, err := virtClient.VirtualMachineInstance(vmi.ObjectMeta.Namespace).USBRedir(vmi.ObjectMeta.Name)
+			Expect(err).To(HaveOccurred())
+			Expect(usbredirVMI).To(BeNil())
+		})
+	})
+
+	Describe("[crit:medium][vendor:cnv-qe@redhat.com][level:component] A VirtualMachineInstance with usbredir support", func() {
+		tests.BeforeAll(func() {
+			virtClient, err = kubecli.GetKubevirtClient()
+			util.PanicOnError(err)
+
+			tests.BeforeTestCleanup()
+			vmi = tests.NewRandomVMI()
+			vmi.Spec.Domain.Devices.ClientPassthrough = &v1.ClientPassthroughDevices{}
+			Expect(virtClient.RestClient().Post().Resource("virtualmachineinstances").Namespace(util.NamespaceTestDefault).Body(vmi).Do(context.Background()).Error()).To(Succeed())
+			tests.WaitForSuccessfulVMIStart(vmi)
+		})
+
+		Context("with an usbredir connection", func() {
+
+			usbredirConnect := func(connStop chan struct{}) {
+				pipeInReader, pipeInWriter := io.Pipe()
+				pipeOutReader, pipeOutWriter := io.Pipe()
+				defer pipeInReader.Close()
+				defer pipeOutReader.Close()
+
+				k8ResChan := make(chan error)
+				readStop := make(chan []byte)
+
+				By("Stablishing communication with usbredir socket from VMI")
+				go func() {
+					defer GinkgoRecover()
+					usbredirVMI, err := virtClient.VirtualMachineInstance(vmi.ObjectMeta.Namespace).USBRedir(vmi.ObjectMeta.Name)
+					if err != nil {
+						k8ResChan <- err
+						return
+					}
+
+					k8ResChan <- usbredirVMI.Stream(kubecli.StreamOptions{
+						In:  pipeInReader,
+						Out: pipeOutWriter,
+					})
+				}()
+
+				By("Exchanging hello message between client and QEMU's usbredir")
+				go func() {
+					defer GinkgoRecover()
+					buf := make([]byte, 1024, 1024)
+
+					// write hello message to remote (VMI)
+					nw, err := pipeInWriter.Write(helloMessageLocal)
+					Expect(err).ToNot(HaveOccurred())
+					Expect(nw).To(Equal(len(helloMessageLocal)))
+
+					// reading hello message from remote (VMI)
+					nr, err := pipeOutReader.Read(buf)
+					if err != nil && err != io.EOF {
+						Expect(err).ToNot(HaveOccurred())
+						return
+					}
+					if nr == 0 && err == io.EOF {
+						log.Log.Info("zero bytes read from usbredir socket.")
```

Comment:
Correct me if I'm wrong, but doesn't this mean that the test failed?
Makes sense to me to report that we read an empty message `readStop <- []byte("")`, that would cause the test to fail since this is not the expected message. This way we don't need to wait for the timeout to expire.